### DISCUSSION TITLE
selectNew with a long field mapping

### DISCUSSION THREAD

**Discussion by helpermethod** [Blazebit/blaze-persistence#1821] [Q&A] [answered]
Hi!

First of all, great project! I've just migrated a project from QueryDSL to Blaze Persistence and it was smooth as butter!

Now, while the migration went smoothly, there was still some weird bug(?):

```
criteriaBuilderFactory
    .create(entityManager, Tuple::class.java)
    .selectNew(ConversionGroup::class.java)
        .with("conversionEvent")
        .with("conversionType")
        .with("count(distinct profilePersonalizationId)")
    .end()
    // ... more stuff, doing a group by by conversionEvent, conversionType etc.
```

`ConversionGroup` is defined as follows

```
data class ConversionGroup(
    /**
     *  A, B, C or D
     */
    val conversionEvent: String,
    val conversionType: String,
    val count: Long,
)
```

When I run the query I'm getting

    java.lang.RuntimeException: No matching constructor

The problem is, that `val count: Long` is basically the same as `long count` in Java, i.e. non-nullable, and Blaze Persistance checks in `ClassObjectBuilder.java` if the Long value is assignable from the long value:

```
    // ClassObjectBuilder.java Line 48 
    if (tuple[i] != null && !paramTypes[i].isAssignableFrom(tuple[i].getClass())) {
```

If I change `val count: Long` to `val count: Long?` it works, because now it's the same as `Long count` in Java.

Is there some way to work around this behaviour, or do I need to use Long? I suspect the problem is the same in Java if I try to map to a long field.

**Comment by helpermethod** [comment]
I suspect the method `ClassObjectBuilder.getConstructor` needs to also take primitive types into account. I guess I could create a PR for this.

**Reply by beikov** [reply, marked as answer]
This would be nice, though I would recommend you to look into Entity-Views instead for modelling this :)

**Reply by helpermethod** [reply]
Yes, Entity-Views would certainly be better, but this is a low-effort migration to get rid of Query DSL, and for that we didn't want to rewrite too much. But it was actually very easy to migrate the Query DSL Queries to Blaze Persistence :).
